File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/2110-공유기 설치/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 200005

int N, C;
int A[NM];

void input() {
	cin >> N >> C;
	for (int i = 1; i <= N; i++) cin >> A[i];
}

bool determination(int D) {
	int cnt = 1, last = A[1];
	for (int i=2;i<=N;i++){
		if (A[i] - last < D) continue;
		last = A[i];
		cnt++;
	}
	return cnt >= C;
}

void pro() {
	sort(A+1,A+1+N);
	int L = 1, R = 1000000000, ans = 0;
	while (L <= R) {
		ll mid = (L + R) / 2;
		if (determination(mid)) {
			ans = mid;
			L = mid + 1;
		}
		else R = mid - 1;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/2110-공유기 설치/solution.py
import sys
si = sys.stdin.readline
n, c = list(map(int, si().split()))
a = []
for i in range(n):
    a.append(int(si()))

def determination(D):
    cnt, last = 1, a[0]
    for i in range(1, n):
        if a[i] - last < D: continue
        last = a[i]
        cnt += 1
    return cnt >= c

a.sort()
l, r, ans = 1, 1000000000, 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        l = mid + 1
    else:
        r = mid - 1

print(ans)